Automattic/jetpack
# Sharing: add a Mastodon sharing button

Fixes #7148

## Proposed changes:

This introduces a new Mastodon sharing button:

<img width="470" alt="image" src="https://user-images.githubusercontent.com/426388/216085445-333f505e-f8ae-4cb2-960f-a9420cc07792.png">
<img width="463" alt="Screenshot 2023-02-13 at 14 56 17" src="https://user-images.githubusercontent.com/426388/218478046-35a684ec-ac5b-452e-b074-a13aee2a1048.png">
<img width="459" alt="Screenshot 2023-02-13 at 14 58 56" src="https://user-images.githubusercontent.com/426388/218478076-3847a103-c1e5-4632-902f-6c00616a5464.png">

### Notes

**Class name**

In the past, we once had to give a unique name to a sharing service's class (`Jetpack_Share_WhatsApp`) to avoid conflicting with an existing plugin that already extended the sharing module. I looked around a bit, but could not find any plugin that does that for Mastodon today; folks who want a sharing button on their site today use the "custom sharing button" option we provide in the sharing settings.

**Mastodon support**

This will only support Mastodon. There are many other ActivityPub clients out there, but Mastodon is the biggest one so far, and is one that does offer a standardized "sharing" interface via the web.

In the future, if other clients become more and more popular, we can consider supporting those as well.

**Custom instance support**

We do, however, want to allow site visitors to select on which instance they want to post. The sharing button will consequently need to support inputing an instance name.

**Publicize**

This adds a sharing button, to allow site visitors to manually share blog posts to a Mastodon instance. For automated posting to Mastodon, we track requests and will post updates in #7790.

Until this is implemented, one can use [the ActivityPub plugin](https://wordpress.org/plugins/activitypub/) to have your site publish all your posts to the Fediverse. You can then Boost those posts from your Mastodon / Fediverse account if you have one.

### To do

- [x] Add a first modal view where one can enter their instance URL (or maybe rather their full Mastodon username, and we'd pull the instance name from that.
- [x] Consider a different default message, maybe pulling in a post's tags as hashtags.
- [x] Add a filter so site owners can customize that default message.
- [x] Ensure the Mastodon sharing button can also be added and managed from the Calypso UI (at `https://wordpress.com/marketing/sharing-buttons/yoursite.com`). This will require a Calypso PR: https://github.com/Automattic/wp-calypso/pull/72876

### Other information:


## Jetpack product discussion

* pewaj1-3x-p2

## Does this pull request change what data or activity we track or use?
* No

## Testing instructions:

You'll want to start with a site that's connected to WordPress.com, or where you've enabled [Jetpack's Offline Mode](https://jetpack.com/support/development-mode/).

* Go to Jetpack > Settings > Sharing
* Enable the sharing buttons
* Go to Settings > Sharing
* Drag a Mastodon sharing button to one of the sharing areas.
    *  You will want with different button styles.
* Check that the button works on the frontend.
